File: internal/app/check.go
package app

import (
	"errors"
	"fmt"
	"os"
	"os/exec"
	"time"

	"github.com/leandronsp/githerb/internal/config"
	"github.com/leandronsp/githerb/internal/review"
)

// ErrCheckKilled is a check that was stopped rather than answered.
var ErrCheckKilled = errors.New("a check that was killed is not a check that failed")

// Check runs what the repository declares against a proposal's head revision
// and writes down what happened.
//
// It runs in a throwaway worktree of that exact commit, not in your working
// tree, so the answer is about the code that would land rather than the code
// you happen to have open. It also means the check can run while you keep
// editing.
type Check struct {
	Proposals review.Proposals
	Config    config.Config
	Root      string
	Author    string
	Now       Clock
}

// Run runs every declared check that has not already answered for this
// revision, and returns the results.
func (c Check) Run(id string) ([]review.Check, error) {
	proposal, err := c.Proposals.Load(review.ProposalID(id))
	if err != nil {
		return nil, err
	}

	head := proposal.Head().SHA()
	already := proposal.Checks()

	var results []review.Check

	for _, name := range c.Config.Required() {
		// A revision that already answered is not asked twice. The commit is
		// the same commit; nothing about it changed while you were away.
		if done, ran := already[name]; ran {
			results = append(results, done)

			continue
		}

		result, err := c.one(name, c.Config.Checks[string(name)], head)
		if err != nil {
			return nil, err
		}

		if err := c.Proposals.Annotate(head, review.CheckRecord(result)); err != nil {
			return nil, err
		}

		results = append(results, result)
	}

	return results, nil
}

func (c Check) one(name review.CheckName, command string, head review.SHA) (review.Check, error) {
	dir, err := os.MkdirTemp("", "githerb-check-")
	if err != nil {
		return review.Check{}, fmt.Errorf("making a worktree: %w", err)
	}

	defer func() {
		//nolint:gosec // G204: the arguments are ours and the paths are ones we made.
		_ = exec.Command("git", "-C", c.Root, "worktree", "remove", "--force", dir).Run()
		_ = os.RemoveAll(dir)
	}()

	//nolint:gosec // G204: the arguments are ours and the revision is a sha we read from a ref.
	add := exec.Command("git", "-C", c.Root, "worktree", "add", "--detach", dir, string(head))

	if out, err := add.CombinedOutput(); err != nil {
		return review.Check{}, fmt.Errorf("worktree for %s: %s: %w", head, out, err)
	}

	began := time.Now()

	// The command comes from the repository, so this runs whatever the
	// repository says. That is the same trust you give a Makefile, and it is
	// why nothing here runs a check for a proposal fetched from someone else
	// without you asking for it.
	run := exec.Command("sh", "-c", command) //nolint:gosec // G204: see above
	run.Dir = dir
	run.Stdout = os.Stdout
	run.Stderr = os.Stderr

	status := review.CheckPassed

	if err := run.Run(); err != nil {
		// A command that died on a signal was killed, by a Ctrl-C or a
		// shutdown, and killing something is not a verdict on it. Recording it
		// as failed would block the revision on an answer nobody gave.
		var exit *exec.ExitError
		if errors.As(err, &exit) && exit.ExitCode() == -1 {
			return review.Check{}, fmt.Errorf("%s was killed: %w", name, ErrCheckKilled)
		}

		status = review.CheckFailed
	}

	result, err := review.NewCheck(name, status, head, int(time.Since(began).Seconds()), c.Author, c.Now())
	if err != nil {
		return review.Check{}, err
	}

	return result, nil
}

File: crates/app/src/check.rs
//! Running what the repository declares against the head revision, and
//! writing down what happened.
//!
//! Every check runs in a throwaway worktree of that exact commit, never in
//! your working tree, so the answer is about the code that would land rather
//! than the code you happen to have open, and you can keep editing while it
//! runs.

use std::io::Write;
use std::path::{Path, PathBuf};
use std::process::Command;
use std::sync::atomic::{AtomicU64, Ordering};
use std::time::Instant;

use review::{Author, Check, CheckName, CheckStatus, ProposalId, Record, Sha, Timestamp};

use crate::config::{Config, FILE};
use crate::error::{Error, Result};
use crate::format;
use crate::store::Store;

/// Makes each worktree directory of one process its own.
static COUNTER: AtomicU64 = AtomicU64::new(0);

/// Run every declared check that has not already answered for this revision,
/// and write each result down as it lands.
///
/// A revision that already answered is not asked twice: the commit is the
/// same commit, and nothing about it changed while you were away. Results are
/// written to `out` in the order they were declared, so a long gate reads as
/// it goes.
///
/// # Errors
///
/// A proposal nobody opened, a worktree git refused to make, or a check that
/// was killed rather than answered.
pub fn check(
    store: &Store,
    config: &Config,
    author: &Author,
    now: Timestamp,
    id: &ProposalId,
    out: &mut dyn Write,
) -> Result<Vec<Check>> {
    let proposal = store.load(id)?;
    let head = proposal.head().sha().clone();
    let already = proposal.checks();

    if config.checks().is_empty() {
        writeln!(out, "no checks declared in {FILE}")?;

        return Ok(Vec::new());
    }

    let mut results = Vec::new();

    for (name, command) in config.checks() {
        let name = CheckName::parse(name)?;

        let result = if let Some(done) = already.iter().find(|done| done.name() == &name) {
            (*done).clone()
        } else {
            let ran = one(store, &name, command, &head, author, now)?;
            store.annotate(&head, &Record::Check(ran.clone()))?;
            ran
        };

        writeln!(out, "{}", format::check_line(&result))?;
        results.push(result);
    }

    Ok(results)
}

/// How many of a set of results said no.
#[must_use]
pub fn refused(results: &[Check]) -> usize {
    results.iter().filter(|result| !result.passed()).count()
}

/// The checks the repository requires, as names the gate can ask for.
///
/// # Errors
///
/// A check declared under a blank name.
pub fn required(config: &Config) -> Result<Vec<CheckName>> {
    config
        .required()
        .into_iter()
        .map(|name| Ok(CheckName::parse(name)?))
        .collect()
}

/// One check, in a worktree of its own.
fn one(
    store: &Store,
    name: &CheckName,
    command: &str,
    head: &Sha,
    author: &Author,
    now: Timestamp,
) -> Result<Check> {
    let dir = scratch();
    std::fs::create_dir_all(&dir)?;
    store.repo().worktree_add(&dir, head.as_str())?;

    let began = Instant::now();
    let status = run(command, &dir);

    remove(store, &dir);

    let status = status?;
    let seconds = u32::try_from(began.elapsed().as_secs()).unwrap_or(u32::MAX);

    Ok(Check::new(
        name.clone(),
        status.ok_or_else(|| Error::CheckKilled(name.clone()))?,
        head.clone(),
        seconds,
        author.clone(),
        now,
    ))
}

/// Run the command the repository declares, and read the exit code as a
/// verdict.
///
/// A command that died on a signal was killed, by a Ctrl-C or a shutdown, and
/// killing something is not a verdict on it: that is `None`, and nothing is
/// recorded. The command comes from the repository, which is the same trust
/// you give a Makefile.
fn run(command: &str, dir: &Path) -> Result<Option<CheckStatus>> {
    let status = Command::new("sh")
        .arg("-c")
        .arg(command)
        .current_dir(dir)
        .status()?;

    Ok(status.code().map(|code| {
        if code == 0 {
            CheckStatus::Passed
        } else {
            CheckStatus::Failed
        }
    }))
}

/// A directory beside the repository, for one check.
fn scratch() -> PathBuf {
    let id = COUNTER.fetch_add(1, Ordering::Relaxed);

    std::env::temp_dir().join(format!("githerb-check-{}-{id}", std::process::id()))
}

/// Take the worktree away whatever the check did to it. Neither failure is
/// worth losing the verdict over: a leaked directory is swept up by the next
/// prune, and git is told about it either way.
fn remove(store: &Store, dir: &Path) {
    let _ignored = store.repo().worktree_remove(dir);
    let _ignored = std::fs::remove_dir_all(dir);
}
